First draft at new explainer (#181)
WICG/portals: README.md

# Portals

> Portals are a proposed new HTML element which enables seamless navigations between pages. In particular, the `<portal>` element enables a page to show another page as an inset, and then "activate" it to perform a seamless transition to a new state, where the formerly-inset page becomes the top-level document.

Reviewer: > becomes the top-level document.

Does this language work if the portal is activating within an iframe?

Author: Portals cannot load inside an iframe.

Reviewer: Ohh I didn't realise that. Hah.

> Portals encompass some of the use cases that iframes currently do, but with better security, privacy, and ergonomics properties. And via the activation operation, they enable new use cases like preloading or navigation transitions.

Author: I wonder if it'd be worth mentioning the "super-links" framing here.

Reviewer: Given the concessions we need to make for cross-origin portals, I'm a little nervous about about saying they're better than iframes. They comes with additional restrictions compared to iframes, so they can be activated without compromising privacy. But ergonomics wise, things like the rendering style will make them less-useful for some of the things folks use iframes for today.

Author: Right, I tried to be explicit it only covers some of the use cases iframes do, with some more details below.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
## Table of contents

- [Example](#example)
- [Use cases](#use-cases)
- [Objectives](#objectives)
- [Details](#details)
  - [Privacy threat model and restrictions](#privacy-threat-model-and-restrictions)
  - [Permissions and policies](#permissions-and-policies)
  - [Rendering](#rendering)
  - [Interactivity](#interactivity)
  - [Accessbility](#accessbility)
  - [Session history, navigation, and bfcache](#session-history-navigation-and-bfcache)
- [Summary of differences between portals and iframes](#summary-of-differences-between-portals-and-iframes)
- [Alternatives considered](#alternatives-considered)
- [Acknowledgments](#acknowledgments)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Example

A document can include a `<portal>` element, which renders the specified URL in a portal context:

```html
<portal id="myPortal" src="https://www.example.com/"></portal>
```

> This is somewhat like an iframe, in that it provides a rendering of the specified document at `https://example.com/`. However, a portal is much more restricted than an iframe: user interaction with it is [limited](#interactivity) to just clicks (like a button or link), and capabilities like storage and cross-origin communication are [prohibited](#privacy-threat-model-and-restrictions) while rendering in the portal context. This allows portals to be used as a more-restricted view onto another document when the full capabilities of an iframe are not needed.

Author: I rephrased this to be more accurate; PTAL!

In exchange for these restrictions, portals gain an additional capability that iframes do not have. A portal can be *activated*, which causes the embedding window to navigate, replacing its document with the portal:

```js
myPortal.activate();
```

> TODO are we still thinking of making this the default click action? If so we should state that here. If it's still under debate we should state that it's under debate and link to an open issue.

Reviewer: It's still under debate.

> At this point, the user will observe that their browser has navigated to `https://www.example.com/` (e.g., via changes to the URL bar contents and back/forward UI). Since `https://example.com/` was already loaded in the portal context, this navigation will occur seamlessly and instantly, without a network round-trip or document re-initialization.

Reviewer: > this navigation will occur seamlessly and instantly, without a network round-trip or document re-initialization.

There may be cases where we need to force a reload. Eg, if a GitHub page is portaled, the storage restrictions will make it render in a logged-out experience, at best. Activating the page in this state may make it look like GitHub is broken/unreliable, so I think GitHub would need to acknowledge the change in storage partitioning to avoid a forced reload.

Author: Hmm. Would the browser force the reload, or would the page add `location.reload()` in its `portalactivate` handler? If the former, how would we know when to force the reload and when not?

Author: Makes sense. I agree we can leave out that level of detail here, although we should have an open issue recording it. Can one of you take point on writing that up in the issue tracker?

Reviewer 2: @jakearchibald is probably better equipped to articulate the problem, since he was the one who pointed it out originally.

Reviewer: On it.

Reviewer: https://github.com/WICG/portals/issues/184

> For more advance use cases, the `https://www.example.com/` document can react to activation, using the `portalactivate` event. It can use this event to adapt itself to its new context, and even to adopt its *predecessor* (the document which previously occupied the tab) into a new portal context.

Reviewer: ```suggestion
For more advanced use cases, the `https://www.example.com/` document can react to activation, using the `portalactivate` event. It can use this event to adapt itself to its new context, and even to adopt its *predecessor* (the document which previously occupied the tab) into a new portal context.
```

```js
window.addEventListener('portalactivate', e => {
  document.body.classList.add('displayed-fully');

>   document.requestStorageAccess();

Author: I threw in some extra lines here because the code seemed uncompelling (and hopefully unnecessary?) if it was basically just "keep me around to work around the lack of bfcache". Suggestions for better stuff welcome.

Author: Done. More ideas still welcome.

- **Navigation transitions**: pre-rendering opens the door for more elaborate transitions, by displaying the portal in some form, animating it (using resizing, translations, etc.) until it occupies the full viewport, then finally activating the portal to perform the instant navigation.

- **Aggregation**: multiple portals on the same page can be used to create more elaborate experiences, where the user chooses which portal to activate. This category of use cases includes cases like a news reader, a shopping site, an infinite scrolling list of articles, etc. By using a portal instead of (or in addition to) a link, the aggregated content has the opportunity to display a preview, and to benefit from pre-rendering and navigation transitions.

  Additionally, by using the ability to adopt the predecessor during the `portalactivate` event, more complicated interactions between the aggregator and the aggregated content can be built, such as retaining a portion of the shopping site or article-list UI in a portal even after navigating to an individual page.

> - **"Better iframe"**: portals encompass some, but not all, of the use cases for iframes. And they do so in a way that is better for users, in terms of security and privacy. They also remove a lot of the legacy baggage and sharp edges that come with iframes, making them easier to use for web developers. So we anticipate many of the current cases for iframes, such as ads, being able to be replaced with portals.

Author: "such as ads" might not be super-accurate here. In general it'd be good to get a better sense of which iframe use cases can be done via portals. (That's one of the TODOs below.)

Reviewer: I'm not sure if you'd use a portal for things that don't need to activate.

Author: I mean, ads need to be navigated to, and activating them seems like a nicer way of navigating? I'm not sure though.

Author: Are there other cases to mention where people might want to use portals instead of iframes? Maybe same-origin ones?

## Objectives

Goals:

- Enable seamless navigations from a page showing a portal, to the portaled page

- Enable seamless navigations between pages of a portal-aware website

- Avoid the characteristics of iframes which have negative impacts on security, privacy, and performance

Non-goals:

- Built-in support for high-level navigation patterns, such as carousels or infinite lists. Portals provide a low-level building block for pre-rendering, which can be combined with the usual tools of HTML for creating navigation pattern UIs.

- Built-in support for portal-specific transition animations. Given that portals are represented by HTML elements, existing CSS mechanisms are enough to allow authors to create compelling navigation transitions.

- Subsume all the use cases of iframes. The use cases for portals overlap with those for iframes, but in exchange for the ability to be activated, portaled pages lose abilities like cross-origin communication, storage, and nontrivial interactivity. As such, portals are not suitable for use cases like embedded widgets.

> - Allowing arbitrary unmodified web pages to be portaled. Pages will need to adapt to work well when they are hosted in a portal.

Reviewer: fwiw, I don't think some-origin pages will need to adapt to being portaled.

Author: For these purposes I'll just add "Cross-origin" to the beginning.

- Somewhere we need to discuss our plans for `X-Frame-Options`, CSP `frame-ancestors`, and portal-specific variants of those? This might be the place for that?

### Rendering

TODO:

- Discuss rendering. Full viewport size at all times? Resizing OK or no? I'm sketchy on the plan here.
- Discuss practices and patterns for authors of portaled sites, e.g. how to adapt on activation.
- Discuss practices and patterns for authors of portaling sites, e.g. how to create a prerender (with a `display: none` portal) or an animated transition
- Maybe this is where we discuss fallback content for non-supporting browsers?

### Interactivity

TODO:

- Discuss how portals are only trivially interactive, like buttons or links. Not like iframes.
- Discuss focus. (Again, like buttons or links.)
- Did we end up adding a default click behavior like links?
- Discuss scrolling, including the problem of scroll handoff and its importance. Note that scroll handoff is a ??? in spec terms but we promise to spec something interoperably implementable, somehow.

> ### Accessbility

Reviewer 3: ```suggestion
### Accessibility
```

- Explicitly mention how authors should respect `prefers-reduced-motion` (and how this might happen automatically if they already turn off all CSS transitions/animations when that media query tests true).

### Session history, navigation, and bfcache

TODO:

- One-sentence reiteration of how activation is a navigation.
- Talk about how session history is tricky.
- Talk about how bfcache is tricky.
- Outline our solution for these, in terms of observable effects for users (not in terms of spec primitives).
- Maybe compare/contrast with iframes. Is it helpful or confusing to talk about how portals are more like popups than iframes? (I.e. they are auxiliary browsing contexts in spec terms.)

## Summary of differences between portals and iframes

TODO:

- Maybe introduce by telling people to think of portals as super-links, not super-iframes. At least from a user's perspective; from the spec and implementer perspective they're more related to iframes, and for web developers they're kind of in between.
- Reiterate everything from the above sections in one convenient place.
- Include minor things as well (like, doesn't appear in `window.frames`).

## Alternatives considered

TODO:

> - A new attribute on iframes (text exists in explainer.md but is very implementer-focused)

Reviewer 3: Text doesn't exist there anymore. Maybe
```suggestion
- A new attribute on iframes ([text exists in explainer.md](https://github.com/WICG/portals/blob/d901563053af7d56a2cafee686ee89d651c49eb7/explainer.md#why-do-we-need-a-new-html-tag-why-not-use-iframe) but is very implementer-focused)
```

- Other (historical?) solutions to prerendering
- Other (historical?) solutions to navigation transitions
- Adding activation ("promotion") to iframes (text exists in explainer.md but is very implementer-focused)

> - Using the fullscreen API (text exists in explainer.md but is very implementer-focused)

Reviewer 3: Ditto here.

- Allowing cross-origin communication and storage

> ## Acknowledgments

Reviewer 3: ```suggestion
## Security and privacy considerations

TODO

### Security/Privacy Questionnaire

This section contains answers to the [W3C TAG Security and Privacy Questionnaire](https://w3ctag.github.io/security-questionnaire/).

1. TODO

## Stakeholder Feedback

* W3CTAG: ?
* Browsers:
  * Safari: ?
  * Firefox: ?
  * Samsung: ?
  * UC: ?
  * Opera: ?
  * Edge: ?
* Web developers: ?

## Acknowledgments
```

From https://w3ctag.github.io/explainers#stakeholder-feedback--opposition and general good practice. It's possible the questionnaire should live in a separate document.

Author: I think a separate document is probably best for the questionnaire. I'll add a skeleton for it.

At this point, the user will observe that their browser has navigated to `https://www.example.com/` (e.g., via changes to the URL bar contents and back/forward UI). Since `https://example.com/` was already loaded in the portal context, this navigation will occur seamlessly and instantly, without a network round-trip or document re-initialization.

For more advanced use cases, the `https://www.example.com/` document can react to activation, using the `portalactivate` event. It can use this event to adapt itself to its new context, and even to adopt its *predecessor* (the document which previously occupied the tab) into a new portal context.

```js
window.addEventListener('portalactivate', e => {
  document.body.classList.add('displayed-fully');
  document.requestStorageAccess().then(() => {
    document.getElementById('user').textContent = localStorage.getItem('current-user');
  });

  let predecessor = e.adoptPredecessor(document);
  console.assert(predecessor instanceof HTMLPortalElement);
  document.body.appendChild(predecessor);
});
```

## Use cases

> _See the ["Key Scenarios"](./key-scenarios.md) document for more detail on each of these, including visualizations._

Reviewer 3: This contradicts the warning at the top of `key-scenarios.md` a bit, but it's probably worth keeping anyway.

> - **Pre-rendering**: by loading a page in a hidden portal, it is possible to pre-render the destination page. Then, activating the portal will instantly display the pre-rendered document in the same browsing context.

Reviewer 3: Is "browsing context" right? Or should this say
```suggestion
- **Pre-rendering**: by loading a page in a hidden portal, it is possible to pre-render the destination page. Then, activating the portal will instantly display the pre-rendered document in the same browser tab.
```
to be more vague about the architecture?

Author: I think both are right, but I prefer your user-facing language suggestion.
